File: MergeSort/MergeSort.c
# include "list.h"
//		printf("\nHERE\n");

Node* half(Node* arr){
	Node* temp = arr;
	Node* center = arr;
	while ( temp->next ){
		if ( temp -> next -> next ){
			temp = temp->next->next;
	       	}
		else{ 
			// since temp->next must exit we will cut its next.
			break ; 
		}
		center = center->next;
		
	}
	return split( center );
}

// Node* _sort(Node* arr){}





Node* sort(Node* arr){
	
	Node* center = half( arr );
	
	printf("\n-------------------------------\n");

	show( center );
	show( arr );

	// If either `arr` or `center` have no next, then the length is one.
	center = sort( center );
	arr = sort( arr );	

	// Take both lists, look at the heads, find the minimum. 
	// Start list using minimum.
	// Repeat
	//Node* head = arr->value < center->value ? arr : center;
	// Node* end;

	return arr;
}

void mergeSort(Node* arr){
	
	

	return;
	
}

void test_half(){

	int _arr[] = {2, 1, 3, 4, 7, 11, 18, 29, 47, 76};
	Node* arr = newList(_arr, 10);
	show( arr );

	Node* center = half( arr );

	show( center );
	show( arr );

}

int main(){
	
	int _arr[] = {2, 1, 3, 4, 7, 11, 18, 29, 47, 76};
	Node* arr = newList(_arr, 10);
	show( arr );

Node* head = sort( arr );
	show( head );
	return 0;
}

File: MergeSort/Py/mergesort.py

def _sort( arr ):
    n = len(arr)
    if n <= 1: return arr

    center = [arr.pop() for k in range(0, n // 2)]
    
    center = _sort( center ) 
    arr = _sort( arr )
    
    out = [ None for item in range( 0, len(center) + len(arr) ) ]

    i = j = k = 0
    n, m = len(arr), len(center)

    while i < n and j < m :
        print(f"{arr = }, {center =}")
        if arr[ i ] < center[ j ]:
            out[k] = arr[i]
            i += 1
        else:
            out[k] = center[j]
            j += 1

        k += 1
    
    while i < n: 
        out[ k ] = arr[ i ]
        k += 1
        i += 1
    while j < m:
        out[ k ] = center[ j ]
        k += 1
        j += 1

    return out

if __name__ == "__main__":

    A = [2,1,3,4,7,11,18,29,47,76]
    B = _sort( A )
    print(B)
